### PR TITLE
[video_player] Passing http headers to file constructor

### PR DESCRIPTION
**Pull request by abdelaziz-mahdy**
when passing m3u8 file to videoplayer as videoplayercontroller.file links insided m3u8 file has to be used with http headers

so i passed it in the  videoplayercontroller.file and changed all things related to it

issue [https://github.com/flutter/flutter/issues/119467](https://github.com/flutter/flutter/issues/119467)

## Pre-launch Checklist

- [ x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [ x] I signed the [CLA].
- [ x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [ x] I listed at least one issue that this PR fixes in the description above.
- [x ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x ] I updated/added relevant documentation (doc comments with `///`).
- [x ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by zezo357** [comment]
now all of the code will allow http headers to passed since it will use one DataSource

i dont know if it will cause problems since i dont have any tests to try it on

**Comment by stuartmorgan** [comment]
Thanks for the contribution! You’ve checked boxes in the PR checklist above that are not reflected in this PR, so I’m assuming this is a work in progress and am marking it as a Draft. Please review the checklist, updating the PR as appropriate, and when the checklist reflects the state of the PR as posted please feel free to mark it as ready for review.

**Comment by zezo357** [comment]
> Thanks for the contribution! You’ve checked boxes in the PR checklist above that are not reflected in this PR, so I’m assuming this is a work in progress and am marking it as a Draft. Please review the checklist, updating the PR as appropriate, and when the checklist reflects the state of the PR as posted please feel free to mark it as ready for review.

Which one is missing since I can't find it 😅

**Comment by stuartmorgan** [comment]
You haven't added any tests, or run the auto-formatter.

**Comment by zezo357** [comment]
> You haven't added any tests, or run the auto-formatter.

Tests can not be added since I don't know how to test these changes using testing code 


Auto formatter I will do it for both java and dart code

**Comment by stuartmorgan** [comment]
Have you read the documentation about different types of plugin tests linked from our contribution guide, and looked at other tests in the plugin?

If you have specific questions about tests we can try to answer them, but we can't accept PRs with no tests.

**Comment by zezo357** [comment]
> Have you read the documentation about different types of plugin tests linked from our contribution guide, and looked at other tests in the plugin?
> 
> If you have specific questions about tests we can try to answer them, but we can't accept PRs with no tests.

i am sorry i didnt dive deep into testing guide

i will check it and implement the needed tests 

thank you.

**Comment by zezo357** [comment]
I added the tests and made the pull request ready for review for a while now 

Should I do something more or just wait?

**Review by camsim99 — CHANGES_REQUESTED** [review]
Did a first pass through to get the review process going!

You need to add Java unit tests for the changes you made in `VideoPlayer.java` since that is the bulk of the change. Also, looks like there are a lot of unrelated formatting changes that can be fixed by running the commands included in the `format` presubmit test error log.

File: packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java
Comment (zezo357) [reply packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java]: can you explain exactly what should be tested? and i will try to do it

Comment (zezo357) [reply packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java]: yes removed isHttp to set http headers in all cases including (m3u8 files too)  


for the test i will try to do it, 
but if there is a repo doing similar test it that will be very helpful ♥

Comment (camsim99) [reply packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java]: This is an example of the test that uses the approach I mentioned above that tests [this constructor](https://github.com/flutter/plugins/blob/09a4a096d1a55f8cb4378e192202ffc5092ee618/packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/CameraZoom.java#L21):
https://github.com/flutter/plugins/blob/09a4a096d1a55f8cb4378e192202ffc5092ee618/packages/camera/camera_android/android/src/test/java/io/flutter/plugins/camera/CameraZoomTest.java#L55-L64

**Comment by GaryQian** [comment]
@zezo357 It seems like you still need to add a test. Let us know when you get around to that and we can move this PR along. Thanks!

**Comment by zezo357** [comment]
> @zezo357 It seems like you still need to add a test. Let us know when you get around to that and we can move this PR along. Thanks!

I failed to add the test, since I don't have much knowledge about it 


If anyone can help in that matter it will be much appreciated

**Comment by camsim99** [comment]
@zezo357 I can work on adding tests to this to help land the change!

**Comment by zezo357** [comment]
> @zezo357 I can work on adding tests to this to help land the change!

Do I need to give you permission? Or you can make a pull request on my fork? 

Let me know if I need to do something to help you

**Comment by ShareZore** [comment]
I was wondering if this problem has been solved yet

**Comment by zezo357** [comment]
> I was wondering if this problem has been solved yet

The java side of tests are still not implemented, I failed to do it sadly and @camsim99 offered to do it but no progress yet I guess

**Comment by ShareZore** [comment]
> 

Android native can do that, but with flutter I don't know how to write it, right

File: packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java
Comment (camsim99) [inline packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java:80]: @zezo357 I am ready to add tests but I have some follow up questions first to make sure I understand the change.

1. Why did you delete the check for `isHttp(...)` and use `new DefaultDataSource.Factory(context, httpDataSourceFactory)` by default versus `httpDataSourceFactory.setDefaultRequestProperties(httpHeaders)`? What happens if the uri is not HTTP?
2. As a follow up to (1), was the fix for passing mu38 files to add the `.setUserAgent(httpHeaders.get(USER_AGENT))` line if `USER_AGENT` is present the HTTP headers?

Comment (zezo357) [reply packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java:80]: 1-If not http all functions will work and it's the main fix since when playing a m3u8 file it's not http 


2- user agent should be added to the http headers If the user wants it (it's needed because some hosts do not allow the default user agent)

**Comment by stuartmorgan** [comment]
Triage checkin: what's the state of this PR? Is it just waiting for tests?

**Comment by zezo357** [comment]
> Triage checkin: what's the state of this PR? Is it just waiting for tests?

Waiting on the java native tests since I couldn't implement those, but the dart tests I already implemented and they are passing

**Comment by camsim99** [comment]
@zezo357 Sorry for the delay on this. Just added tests and made some minor changes. Let me know what you think!

**Comment by zezo357** [comment]
Will check it thoroughly tomorrow and test it

But for now I see it looks good

**Comment by zezo357** [comment]
looks good to me.

**Comment by camsim99** [comment]
@zezo357 Can you fix the failing Dart test? Then we'll need a second review.

**Comment by zezo357** [comment]
> @zezo357 Can you fix the failing Dart test? Then we'll need a second review.

will check them

**Comment by zezo357** [comment]
I fixed the dart test related to my pull request there is another one that fails 
`file with special characters`
with error 
```
Invalid argument(s): Illegal character in path
dart:core                                      new _Uri.file
package:video_player/video_player.dart 254:26  new VideoPlayerController.file
test\video_player_test.dart 352:35             main.<fn>.<fn>.<fn>
```

**Comment by camsim99** [comment]
@zezo357 I don't see that failure, but there is a formatting error, but you can run the command suggested to fix it https://github.com/flutter/plugins/pull/6213/checks?check_run_id=10935748930

**Comment by zezo357** [comment]
> @zezo357 I don't see that failure, but there is a formatting error, but you can run the command suggested to fix it https://github.com/flutter/plugins/pull/6213/checks?check_run_id=10935748930

fixed formatting

File: packages/video_player/video_player/CHANGELOG.md
Comment (stuartmorgan) [inline packages/video_player/video_player/CHANGELOG.md:4]: Adds

Please see https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style

File: packages/video_player/video_player/lib/video_player.dart
Comment (stuartmorgan) [inline packages/video_player/video_player/lib/video_player.dart:233]: This change is confusing; it makes it sound like the user agent is separate from the headers, but there is no such option. It's not at all clear to a reader how this comment maps to the parameters.

File: packages/video_player/video_player_android/CHANGELOG.md
Comment (stuartmorgan) [inline packages/video_player/video_player_android/CHANGELOG.md:17]: Same style issue here. Also, this was merged incorrectly at some point, as it's adding a duplicate of an existing version now.

File: packages/video_player/video_player/lib/video_player.dart
Comment (stuartmorgan) [inline packages/video_player/video_player/lib/video_player.dart:248]: This comment needs to be updated to explain what `httpHeaders` does in this context, since it's non-obvious.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
> - [x] I listed at least one issue that this PR fixes in the description above.

Please add the issue link to the PR description.

Is this problem actually Android-specific, or does it apply to iOS as well? It's fine to fix it only for Android for now if you don't have iOS experience, but if it's broken there as well we should update the relevant issue accordingly.

**Comment by zezo357** [comment]
> > * [x]  I listed at least one issue that this PR fixes in the description above.
> 
> Please add the issue link to the PR description.
> 
> Is this problem actually Android-specific, or does it apply to iOS as well? It's fine to fix it only for Android for now if you don't have iOS experience, but if it's broken there as well we should update the relevant issue accordingly.

i didnt test on ios, but the issue was for me on android. and i did describe the issue on the pull request should i open an issue?

**Comment by stuartmorgan** [comment]
Yes, that checklist item is about having an issue filed clearly describing the problem, including reproduction steps, so there's a place to reference details about the issue being fixed.

**Comment by zezo357** [comment]
> Yes, that checklist item is about having an issue filed clearly describing the problem, including reproduction steps, so there's a place to reference details about the issue being fixed.

ok i will open one and try to make the steps to reproduce it.

**Comment by zezo357** [comment]
i added the issue in the description with example repo due to using packages to retrevie new urls

since the urls lifetime is short

**Comment by stuartmorgan-g** [comment]
We've just completed the migration of the plugin code to the flutter/packages repository, as described in https://flutter.dev/go/flutter-plugins-repo-migration, and this repository is now being archived. Unfortunately that means that all in-progress PRs here must be moved to flutter/packages.

Please see [our instructions](https://github.com/flutter/flutter/wiki/Migrating-Plugins-repository-PRs-to-Packages) for an explanation of how to move your PR, and if you have any issues moving your PR please don't hesitate to reach out in the #hackers-ecosystem channel in [Discord](https://github.com/flutter/flutter/wiki/Chat).

Our apologies that your PR was caught in this one-time transition. We're aware that it's disruptive in the short term, and appreciate your help in getting us to a better long-term state!

**Pull request closed without merging**
